language_model.py

```diff
@@ -0,0 +1,113 @@
+# Lint as: python3
+"""Trains an nltk language model."""
+
+import pickle
+from typing import List, Tuple
+from nltk.lm.preprocessing import padded_everygram_pipeline
+from nltk.lm import Laplace
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('string_to_normalize', None, 'the string to normalize')
+flags.DEFINE_string('language', None, 'the language to normalize')
+flags.DEFINE_string('data_source', None, 'data source to preprocess')
+flags.DEFINE_string('pass_valid', "token", 'pass only valid tokens or sentences')
+flags.DEFINE_string('experiment', None, 'the normalization experiment to run')
+
+
+def main(argv):
+    """Trains an nltk language model.
+
+    Loads in a file of normalized text, partitions it into a train partition
+    (3/4 of data) and a test partition (last 1/4 of data). Uses Laplace
+    smoothing for unseen ngrams.
+    """
+    if len(argv) > 1:
+        raise app.UsageError("Too many command-line arguments.")
+
+    normalized_data = load_normalized_data(FLAGS.language,
```

Comment:
Could you explain a little bit more what the flags here are doing?  How I imagine it is that the normalization process is entirely separate from the training process, where the rewrites that we want to experiment with are added into the language specific configurations, then the normalizer is run on the data, then we feed the normalized data into the training pipeline.

Reply:
The idea here is that when running `normalizer.py`, you'll pass in flags for the language, data source, and the specific normalization experiment. The flags will determine the name of the output files and what output directory they go to. This will give us something like `language=af_datasource=oscar_passvalid=token_normalized.p`, where these files will end up in different directories within `output/` for each different normalization experiment.

In `language_model.py`, we want to load in the file generated by `normalizer.py`. I'm currently using the same flags here to make loading in the filename consistent and modular. Even though they have the same names as the flags in `normalizer.py`, they are essentially separate. However, if this adds too much complication, we could easily just switch the flags out here for a `sys.argv[1]` sort of thing and just load the filename directly as an argument.

Reply:
Ahhh I see, that makes sense!  Do we want to be able to pass in multiple training files to language_model.py though?

Reply:
So to run the normalization step, we can enter something like `bazel build normalizer && bazel-bin/normalizer --language=yo --data_source=ud --experiment=no-diacritics`, and the normalizer will generate a file called `language=yo_data_source=ud_passvalid=token.normalized.p` in the directory `output/no-diacritics/`.

Then, to train and test the language model on that output, we can run `python3 language_model.py --language=yo --data_source=ud --experiment=no-diacritics`, and the lm will load the correct file, partition it, train the lm with the training partition, and evaluate the perplexity of the test partition.

Does that make sense? My thoughts were to keep it sorta modular, so for a given experiment I could just quickly change e.g. the `language` flag, but it might be just as easy and in fact less complicated to load the whole filename as a command line argument rather than do all this flag stuff here.

Reply:
That definitely makes sense, and as discussed this morning, it sounds like we do want to keep the datasets separate for some of it, but I think we also want to train a single model using all of the available normalized data, correct?  Will this still allow for that?

Reply:
It currently does not. Right now it only reads in one single file and trains the model with it. I can definitely retool it a bit to allow reading in multiple files.

Reply:
Ok, thanks Andrew!  Let me know when that's done and I'll take another look!

Reply:
@crewee I modified the `data_source` flag to take a (space-separated) list instead of a string and changed the `load_normalized_data` method to allow multiple files. I also made sure to shuffle the data after loading the different files, so the later files don't always end up in the testing partition.

This will let us train/test a language model using all the data files for a given language, but currently does not support training/testing multiple languages at the same time. That seems pretty reasonable to me though. What do you think?

Reply:
Great!  Thanks!